spring-integration-jdbc/src/test/java/org/springframework/integration/jdbc/lock/JdbcLockRegistryTests.java

```diff
@@ -0,0 +1,286 @@
+			Lock lock = this.registry.obtain("foo");
+			lock.lockInterruptibly();
+			try {
+				assertEquals(1, TestUtils.getPropertyValue(this.registry, "locks", Map.class).size());
+			}
+			finally {
+				lock.unlock();
+			}
+		}
+	}
+
+	@Test
+	public void testReentrantLock() throws Exception {
+		for (int i = 0; i < 10; i++) {
+			Lock lock1 = this.registry.obtain("foo");
+			lock1.lock();
+			try {
+				Lock lock2 = this.registry.obtain("foo");
+				assertSame(lock1, lock2);
+				lock2.lock();
+				lock2.unlock();
+			}
+			finally {
+				lock1.unlock();
+			}
+		}
+	}
+
+	@Test
+	public void testReentrantLockInterruptibly() throws Exception {
+		for (int i = 0; i < 10; i++) {
+			Lock lock1 = this.registry.obtain("foo");
+			lock1.lockInterruptibly();
+			try {
+				Lock lock2 = this.registry.obtain("foo");
+				assertSame(lock1, lock2);
+				lock2.lockInterruptibly();
+				lock2.unlock();
+			}
+			finally {
+				lock1.unlock();
+			}
+		}
+	}
+
+	@Test
+	public void testTwoLocks() throws Exception {
+		for (int i = 0; i < 10; i++) {
+			Lock lock1 = this.registry.obtain("foo");
+			lock1.lockInterruptibly();
+			try {
+				Lock lock2 = this.registry.obtain("bar");
+				assertNotSame(lock1, lock2);
+				lock2.lockInterruptibly();
+				lock2.unlock();
+			}
+			finally {
+				lock1.unlock();
+			}
+		}
+	}
+
+	@Test
+	public void testTwoThreadsSecondFailsToGetLock() throws Exception {
+		final Lock lock1 = this.registry.obtain("foo");
+		lock1.lockInterruptibly();
+		final AtomicBoolean locked = new AtomicBoolean();
+		final CountDownLatch latch = new CountDownLatch(1);
+		Future<Object> result = Executors.newSingleThreadExecutor().submit(new Callable<Object>() {
+
+			@Override
+			public Object call() throws Exception {
+				Lock lock2 = JdbcLockRegistryTests.this.registry.obtain("foo");
+				locked.set(lock2.tryLock(200, TimeUnit.MILLISECONDS));
+				latch.countDown();
+				try {
+					lock2.unlock();
+				}
+				catch (Exception e) {
+					return e;
+				}
+				return null;
+			}
+		});
+		assertTrue(latch.await(10, TimeUnit.SECONDS));
+		assertFalse(locked.get());
+		lock1.unlock();
+		Object ise = result.get(10, TimeUnit.SECONDS);
+		assertThat(ise, instanceOf(IllegalMonitorStateException.class));
+		assertThat(((Exception) ise).getMessage(), containsString("You do not own"));
+	}
+
+	@Test
+	public void testTwoThreads() throws Exception {
+		final Lock lock1 = this.registry.obtain("foo");
+		final AtomicBoolean locked = new AtomicBoolean();
+		final CountDownLatch latch1 = new CountDownLatch(1);
+		final CountDownLatch latch2 = new CountDownLatch(1);
+		final CountDownLatch latch3 = new CountDownLatch(1);
+		lock1.lockInterruptibly();
+		Executors.newSingleThreadExecutor().execute(new Runnable() {
+
+			@Override
+			public void run() {
+				Lock lock2 = JdbcLockRegistryTests.this.registry.obtain("foo");
+				try {
+					latch1.countDown();
+					lock2.lockInterruptibly();
+					latch2.await(10, TimeUnit.SECONDS);
+					locked.set(true);
```

Comment:
This looks hacky and doesn't demonstrate a general premise of `Lock`.
The second thread should 
> becomes disabled for thread scheduling purposes and lies dormant until the lock has been acquired.

So, that `latch2.await()` is like a trick to make test-case happy, where we really should wait here on the `lock2.lockInterruptibly()` for the `lock1.unlock()` in the main thread.

I won't mind if that is the same in the `Zookeeper` tests, but let's consider to fix it in the separate issue. :smile: 

What is the best scenario to test in the exclusive access mode?
Write to `Queue` some sequence from one (main) thread, than unlock and continue to write that sequence after acquiring lock in the another thread, and `Queue.poll()` finally to be sure that we are fine?